pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,236 @@
+// the following scenario might happen:
+//  1. A new machine joins the cluster, initially as a learner
+//  2. This controller adds a hook for the new machine
+//  3. A user deletes the machine
+//  4. The other controller misses 3 and decides to promote to the voting member, since this is a network call it takes arbitrary long to finish
+//  5. This controller removes the hook because the machine was deleted and the etcd member wasn't promoted to a voting member
+//  6. The network call from 4 finishes and the member is promoted
+//  7. At some point, the machine is removed from the etcd cluster, and we lose a member in an uncontrolled way
+func NewMachineDeletionHooksController(
+	isMachineAPIFunctional func() (bool, error),
+	masterMachineSelector labels.Selector, masterNodeSelector labels.Selector,
+	machineClient machinev1beta1client.MachineInterface,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	networkInformer configv1informers.NetworkInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &machineDeletionHooksController{
+		machineClient:                     machineClient,
+		isMachineAPIFunctionalFn:          isMachineAPIFunctional,
+		masterMachineSelector:             masterMachineSelector,
+		masterNodeSelector:                masterNodeSelector,
+		masterMachineLister:               masterMachineLister,
+		masterNodeLister:                  masterNodeLister,
+		networkLister:                     networkInformer.Lister(),
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithInformers(
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).
+		ResyncEvery(7*time.Minute). // an arbitrary number, not too high not too low
+		ToController("MachineDeletionHooksController", eventRecorder.WithComponentSuffix("cluster-machine-deletion-hooks-controller"))
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, _ factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	// step 1: add hooks for voting members
+	if err := c.addDeletionHookToMasterMachines(ctx); err != nil {
+		return err
+	}
+	// step 2: remove hooks from machines pending deletion and without running etcd member
+	//         before make sure the etcd cluster is healthy and voting members == controlPlane.replicas + 1
+	//
+	//         TODO: step 2 requires updating the ClusterMemberController to not re-add removed members
```

Comment:
@deads2k @hasbro17 how about this? does it also make sense?

Reply:
> @deads2k @hasbro17 how about this? does it also make sense?

This would be represented as a check in clustermembercontroller to avoid adding members that correspond to machines that are pending deletion?